File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        Stack<Integer> s = new Stack<>();
        
        int n = heights.length;
        int max_area = Integer.MIN_VALUE;
        for(int i = 0; i <= n - 1; i++)
        {
            if(s.isEmpty() || heights[i] >= heights[s.peek()])
            {
                s.push(i);
            }
            else
            {
                while(!s.isEmpty() && heights[s.peek()] > heights[i])
                {
                    int processedIndex = s.peek();
                    s.pop();
                    if(s.isEmpty())
                        max_area = Math.max(max_area, i * heights[processedIndex]);
                    else
                        max_area = Math.max(max_area, (i - s.peek() - 1) * heights[processedIndex]);
                }
                s.push(i);
            }
        }
        
        while(!s.isEmpty())
        {
            int processedIndex = s.peek();
            s.pop();
            if(s.isEmpty())
                max_area = Math.max(max_area, n * heights[processedIndex]);
            else
                max_area = Math.max(max_area, (n - s.peek() - 1) * heights[processedIndex]);
        }
        
        
        return max_area;
    }
}

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        stack<int> s;
        int n = heights.size();
        int maxArea = 0;
        for(int i = 0; i<=n-1; i++)
        {
            if(s.empty() || heights[i]>= heights[s.top()])
                s.push(i);
            else
            {
                while(!s.empty() && heights[i] < heights[s.top()])
                {
                    int height = heights[s.top()];
                    int area = 0;
                    s.pop();
                    if(!s.empty())
                    {
                        area = height * (i - s.top() - 1);
                    }
                    else
                        area = height * i;
                    maxArea = max( maxArea, area);
                }
                s.push(i);
            }
        }
        
        while(!s.empty())
        {
            int height = heights[s.top()];
            int area = 0;
            s.pop();
            if(!s.empty())
            {
                area = height * (n - s.top() - 1);
            }
            else
                area = height * n;
            maxArea = max( maxArea, area);
        }
        
        return maxArea;
        
    }
};